src/data/sym/card.lean

```diff
@@ -36,9 +35,185 @@ Prove the general case of stars and bars.
+      convert multiset.map_congr _,
+      { rw multiset.map_id, },
+      rintros ⟨g, hg⟩ h',
+      split_ifs with h'',
+      { simp only [fin.coe_mk] at h'',
+        subst g,
+        exact false.elim (h h'), },
+      { refl } },
+  end,
+  right_inv := begin
+    rintro (x|x),
+    { cases x with x hx,
+      rw [decode, encode],
+      split_ifs,
+      { obtain ⟨y, v, b⟩ := multiset.mem_map.mp h,
+        rw [fin.last, subtype.mk_eq_mk] at b,
+        have := y.property,
+        rw b at this,
+        exact nat.lt_asymm this this, },
+      { simp only [multiset.map_map, function.comp],
+        simp only [fin.val_eq_coe, subtype.mk_eq_mk, multiset.map_map, fin.coe_mk],
+        convert multiset.map_congr _,
+        { rw multiset.map_id, },
+        rintros ⟨g, hg⟩ h',
+        split_ifs with h'',
+        { simp only [fin.coe_mk] at h'',
+          subst g,
+          exact (nat.lt_asymm hg hg).elim, },
+        { refl } } },
+    { rw [decode, encode],
+      split_ifs,
+      { simp, },
+      { apply h,
+        apply multiset.mem_cons_self, } }
+  end }
+
+lemma multichoose1_rec (n k : ℕ) :
+  multichoose1 n.succ k.succ = multichoose1 n k.succ + multichoose1 n.succ k :=
+begin
+  simp only [multichoose1, fintype.card_sum.symm],
+  exact fintype.card_congr (equivalent n k),
+end
+
+lemma multichoose2_rec (n k : ℕ) :
+  multichoose2 n.succ k.succ = multichoose2 n k.succ + multichoose2 n.succ k :=
+begin
+  simp only [multichoose2, nat.add_succ, tsub_zero, nat.succ_sub_succ_eq_sub, nat.succ_add_sub_one,
+    nat.succ_add, nat.choose_succ_succ, nat.add_comm],
+end
+
+lemma multichoose1_eq_multichoose2 : ∀ (n k : ℕ), multichoose1 n k = multichoose2 n k
+| 0 0 := begin
+  simp only [multichoose1, multichoose2],
+  norm_num,
+  dec_trivial,
+end
+| 0 (k + 1) := begin
+  simp only [multichoose1, multichoose2],
+  norm_num,
+  have no_inhabitants : sym (fin 0) k.succ → false := begin
+    intro h,
+    rw sym at h,
+    have w := @multiset.exists_mem_of_ne_zero _ h.val begin
+      intro y,
+      have z := h.property,
+      rw y at z,
+      norm_num at z,
+    end,
+    rcases w with ⟨⟨v, w⟩, q⟩,
+    norm_num at w,
+  end,
+  exact (@fintype.card_eq_zero_iff (sym (fin 0) k.succ) _).mpr ⟨no_inhabitants⟩,
+end
+| (n + 1) 0 := begin
+  simp only [multichoose1, multichoose2],
+  norm_num,
+  dec_trivial,
+end
+| (n + 1) (k + 1) := begin
+  simp only [multichoose1_rec, multichoose2_rec, multichoose1_eq_multichoose2 n k.succ,
+    multichoose1_eq_multichoose2 n.succ k],
+end
+
 open finset fintype
 
 namespace sym2
+
+lemma stars_and_bars {α : Type*} [decidable_eq α] [fintype α] (n : ℕ) :
+  fintype.card (sym α n) = (fintype.card α + n - 1).choose n :=
+begin
+  have start := multichoose1_eq_multichoose2 (fintype.card α) n,
+  simp only [multichoose1, multichoose2] at start,
+  rw start.symm,
+  have bundle := (@fintype.equiv_fin_of_card_eq α _ (fintype.card α) rfl),
+  apply fintype.card_congr,
+  refine ⟨_, _, _, _⟩,
```

Comment:
Can to extract the goal at this point to its own definition? Maybe `sym2.map_equiv` would be a good name.